### PR TITLE
06 - Domain register contract with tests

### PR DESCRIPTION
**Pull request by cesarion161**
Done all the functionality with tests and docs

### CODE REVIEW

File: 6/contracts/DomainRegister.sol [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import "./DomainValidation.sol";
+
+/**
+ * @title Domain Registration Contract
+ * @author Oleksandr Radchenko
+ * @dev Stores domain registrations
+ */
+contract DomainRegister {
+
+    struct Domain {
```
Comment (liscody) [inline 6/contracts/DomainRegister.sol:13]: Over-logic. The logic is too complex. The task did not include the need to check timestamps or create Structs.

- 1 point

File: 6/contracts/DomainRegister.sol [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        string name;
+        uint256 registrationTime;
+    }
+
+    address payable public owner;
```
Comment (liscody) [inline 6/contracts/DomainRegister.sol:18]: What is heppened if need to change owner address?
Uncompleted logic
- 1 point

File: 6/contracts/DomainRegister.sol [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    uint256 public registrationFee = 0.005 ether;
+    mapping(address => Domain[]) private controllerDomains;
+    Domain[] private allDomains;
```
Comment (liscody) [inline 6/contracts/DomainRegister.sol:21]: Double store. 
You strore same domain in two places.

- 3 points

File: 6/contracts/DomainRegister.sol [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "Only the owner can call this function");
+        _;
+    }
+
+    modifier paysFee() {
+        require(msg.value >= registrationFee, "Insufficient fee");
+        _;
+    }
```
Comment (liscody) [inline 6/contracts/DomainRegister.sol:38]: Modifier to check if the caller has paid the registration fee
    but it used only once in the contract
    so it can be replaced with a simple require statement
    // - 1 point

File: 6/contracts/DomainRegister.sol [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    /**
+     * @notice Updates the registration fee for domain registration
+     * @dev This function can only be called by the contract owner. Emits a `FeeUpdated` event upon success
+     * @param newFee The new registration fee in wei
+     */
+    function updateRegistrationFee(uint256 newFee) external onlyOwner {
+        registrationFee = newFee;
+        emit FeeUpdated(newFee);
+    }
+
```
Comment (liscody) [inline 6/contracts/DomainRegister.sol:82]: Should validate newFee parameter

// - 1 point

File: 6/contracts/DomainRegister.sol [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    /**
+     * @notice Withdraws all funds collected by the contract to the owner's address
+     * @dev This function can only be called by the contract owner
+     */
+    function withdraw() external onlyOwner {
+        uint256 balance = address(this).balance;
+        (bool sent,) = owner.call{value: balance}("");
+        require(sent, "Failed to send fee to the owner");
+    }
```
Comment (liscody) [inline 6/contracts/DomainRegister.sol:91]: do not contraine possible to withdraw all funds for another address

// - 1 point

File: 6/contracts/DomainRegister.sol [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+pragma solidity ^0.8.24;
+
+import "./DomainValidation.sol";
+
+/**
+ * @title Domain Registration Contract
+ * @author Oleksandr Radchenko
+ * @dev Stores domain registrations
+ */
+contract DomainRegister {
```
Comment (liscody) [inline 6/contracts/DomainRegister.sol:11]: contract do not have a posibilyty to understand how many domains registered

// - 1 point

File: 6/contracts/DomainRegister.sol (hunk 5 above)
Comment (cesarion161) [reply 6/contracts/DomainRegister.sol:82]: @liscody how validate? no requirements for that

File: 6/contracts/DomainRegister.sol (hunk 2 above)
Comment (cesarion161) [reply 6/contracts/DomainRegister.sol:18]: No requirements for that

File: 6/contracts/DomainRegister.sol (hunk 6 above)
Comment (cesarion161) [reply 6/contracts/DomainRegister.sol:91]: No requirements in the task for that @liscody 

> do not contraine possible to withdraw all funds for another address
> 
> // - 1 point

File: 6/contracts/DomainRegister.sol (hunk 7 above)
Comment (cesarion161) [reply 6/contracts/DomainRegister.sol:11]: The task was to collect metrics about the amount of domains, but not to return it from the contract @liscody

File: 6/test/DomainRegisterTest.ts [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+const INSUFFICIENT_FEE = ethers.parseEther("0.001");
+const NEW_FEE = ethers.parseEther("0.01");
+
+async function deploy() {
+    const [owner, nonOwner1, nonOwner2] = await ethers.getSigners();
+    const domainRegister = await hre.ethers.deployContract("DomainRegister");
+    return {domainRegister, owner, nonOwner1, nonOwner2};
+}
+
+describe("Domain registration", function () {
```
Comment (liscody) [inline 6/test/DomainRegisterTest.ts:17]: In general tests should be independent of each other,  and should cover all edge cases separately
f.e.
one describe block for domain registration contain check for error and success cases with event  should also check storage state after each operation
and transfer eth 

// - 15 points

File: 6/contracts/DomainRegister.sol [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    string[] private allDomains;
+
+    event DomainRegistered(address indexed controller, string domain, uint256 timestamp, uint256 totalRegistered);
+    event FeeUpdated(uint256 newFee);
+
+    constructor() {
+        owner = payable(msg.sender);
+    }
+
+    modifier onlyOwner() {
```
Comment (IhorSolDevRD) [inline 6/contracts/DomainRegister.sol:22]: line 22
NatSpec: Modifier should be declared before the function
// - 1 point

File: 6/contracts/DomainRegister.sol [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        require(msg.sender == owner, "Only the owner can call this function");
+        _;
+    }
+
+    /**
+     * @dev Registers a new domain
+     * @param domain The domain name to register
+     * Emits a {DomainRegistered} event
+     */
+    function registerDomain(string memory domain) external payable {
```
Comment (IhorSolDevRD) [inline 6/contracts/DomainRegister.sol:32]: line 32 
function permit registry same domaine name for different owner
no income parameter validation
// - 3 point

File: 6/contracts/DomainRegister.sol [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        allDomains.push(domain);
+
+        emit DomainRegistered(msg.sender, domain, block.timestamp, allDomains.length);
+    }
+
+    /**
+     * @dev Retrieves all registered domains
+     * @return string[] domain names
+     */
+    function getDomains() external view returns (string[] memory) {
```
Comment (IhorSolDevRD) [inline 6/contracts/DomainRegister.sol:43]: line 43
return array with small length passed well but when the length is big the function will fail
and it will be difficult tu manipulate the array in further steps 
// - 1 point

File: 6/contracts/DomainRegister.sol [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        uint256 balance = address(this).balance;
+        (bool sent,) = to.call{value: balance}("");
+        require(sent, "Failed to send fee to the owner");
+    }
+
+    /**
+     * @return number of all registered domains
+     * @dev This function can only be called by the contract owner
+     */
+    function getRegisteredDomainCount() external onlyOwner returns (uint) {
```
Comment (IhorSolDevRD) [inline 6/contracts/DomainRegister.sol:72]: line 72
 Function state mutability can be restricted to view
// - 1 point

File: 6/test/DomainRegisterTest.ts [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+    it("Should allow registration of a valid domain", async function () {
+        const {domainRegister, owner} = await loadFixture(deploy);
```
Comment (IhorSolDevRD) [inline 6/test/DomainRegisterTest.ts:20]: // tests 
line 20 
you can move the the line in before the function and do not load each time
// - 1 point
